### PR TITLE
Integrate mempool with shard chain

### PR DESCRIPTION
**Pull request by bowenwang1996**
- Refactor shard chain to put `transaction_address` and `transaction_results` into `BlockChain`. 
- Rename `ShardBlockChain` to `ShardClient`
- Integrate mempool with ShardClient.

Todo: figure out the details for beacon chain transaction/receipt and beacon chain mempool.

### CODE REVIEW

**Comment by bowenwang1996** [comment]
`transaction_address` and `transaction_results` are moved to `BlockChain` because both beacon chain and shard chain need them. It is not intended to break encapsulation of shard-specific code, but rather to honor it. Beacon chain potentially also need a mempool, and that part hasn't been fully figured out yet.

**Comment by nearmax** [comment]
> `transaction_address` and `transaction_results` are moved to `BlockChain` because both beacon chain and shard chain need them.

Why beacon chain needs them? We have discussed multiple times that beacon chain is not running the runtime, does not have a state, and does not have transactions.

>Beacon chain potentially also need a mempool, and that part hasn't been fully figured out yet.

If it is not figured out yet then let's not change the code. Let's change the code only we figured out what we want to do.

**Comment by nearmax** [comment]
@bowenwang1996 Could you point me to the line that does the following:
> * Integrate mempool with ShardClient.

So far I see that it only undoes #583 and renames `ShardBlockChain` to `ShardClient`

**Comment by bowenwang1996** [comment]
Staking transactions will be crossed-linked to beacon chain in the form of receipts and I think we would like to keep track of them as well.

**Comment by nearmax** [comment]
Staking transactions are submitted to shard chain and end up in shard block, which are in turn cross-linked on the beacon chain. Beacon chain authorities are going to discover shard blocks through the same mechanism authorities use to join and keep up with the shard chains.

**Comment by bowenwang1996** [comment]
https://github.com/nearprotocol/nearcore/blob/f89696f0c5ac0fb961be1489278817c8f8febe08/node/shard/src/lib.rs#L51. Also I don't see where I undid changes made in #583

**Comment by nearmax** [comment]
> [nearcore/node/shard/src/lib.rs](https://github.com/nearprotocol/nearcore/blob/f89696f0c5ac0fb961be1489278817c8f8febe08/node/shard/src/lib.rs#L51)
> Line 51 in [f89696f](/nearprotocol/nearcore/commit/f89696f0c5ac0fb961be1489278817c8f8febe08)
>  pool: Pool, 

This only adds the pool field to the ShardBlockChain, but it does not use it. Calling it integration is a stretch.

> . Also I don't see where I undid changes made in #583

This PR does three things:
* Adds one field `pool` to `ShardBlockChain`, but does not use it. It requires more work to be called integration of mempool into `ShardBlockChain`;
* Renames `ShardBlockChain` to `ShardClient`;
* Moves a bunch of code from `ShardBlockChain` to `BlockChain` and from `ShardChainStorage` to `BlockChainStorage`. #583 Introduced separation of Storage into `ShardChainStorage`, `BeaconChainStorage`, and `BlockChainStorage`. This PR undoes that separation.

In general, regarding our goals. Our goal with the mempool for alphanet is to allow efficient transaction exchange. Our plan for alphanet is to have one consensus (not two for shard and beacon chain), and one shard. Meaning beacon chain will not need mempool. What we do after alphanet is not figured yet, as you said, I suggest we do not modify code just in case.

**Comment by bowenwang1996** [comment]
Will add some code later that uses pool. I don't think moving `transaction_address` and `transaction_results` undid the separation, as the abstractions still remain intact. @SkidanovAlex told me that we will need to store transactions and receipts on beacon chain and hence the change. The change to shard chain storage does not introduce any new features or change existing functionalities.  Therefore I don't think it will break any existing code.

**Comment by nearmax** [comment]
> I don't think moving `transaction_address` and `transaction_results` undid the separation, as the abstractions still remain intact.

This breaks encapsulation of `BlockChain` and `BlockChainStorage`, what abstractions are you talking about when saying they remain intact?

>@SkidanovAlex told me that we will need to store transactions and receipts on beacon chain and hence the change.

I am sure what he meant is that transactions are going to be stored as part of shard blocks, and shard blocks are cross-linked to the beacon chain.

> The change to shard chain storage does not introduce any new features or change existing functionalities.

Sorry, I do not understand what you were trying to say with this sentence.

>Therefore I don't think it will break any existing code.

Just because the code compiles does not mean the change is not breaking architecture.

**Comment by nearmax** [comment]
To summarize this conversation (since it seems like we are going in circles), I do not support this PR, because:
* It does not bring us closer to the goal of alphanet. (Transactions on beacon chain are far beyond the scope of alphanet);
* It does not justify why pool cannot depend on `ShardBlockChain` or `ShardChainStorage`;
* It breaks encapsulation of `BlockChain` and `BlockChainStorage` that was justified in #583 ;
* It does forward optimization saying that we might potentially need transactions on beacon chain at some point in the future;
* It does not introduce new features. Adding mempool as an unused field to ShardBlockChain does not qualify as integration of mempool.

**Comment by bowenwang1996** [comment]
I reverted my change to storage and changed mempool so that now it depends on `ShardChainStorage` rather than `BlockChain`

File: core/mempool/src/lib.rs [hunk 1]
```diff
@@ -1,143 +1,142 @@
...
+const POISONED_LOCK_ERR: &str = "The lock was poisoned.";
 
 /// mempool that stores transactions and receipts for a chain
-struct Pool {
-    transactions: HashSet<SignedTransaction>,
-    receipts: HashSet<ReceiptBlock>,
-    // TODO: make it work for beacon chain as well
-    chain: Arc<ShardBlockChain>,
+pub struct Pool {
+    transactions: RwLock<HashSet<SignedTransaction>>,
```
Comment (ilblackdragon) [inline core/mempool/src/lib.rs]: hm, we currently have lock on the whole client, what these locks adding to that?

Comment (bowenwang1996) [reply core/mempool/src/lib.rs]: Oh I didn't realize that :(

Comment (bowenwang1996) [reply core/mempool/src/lib.rs]: Actually in network we don't have lock on client

**Comment by nearmax** [comment]
@bowenwang1996 can we also rename `BeaconBlockChain` to `BeaconClient` for consistency?

**Comment by bowenwang1996** [comment]
Sure

**Pull request merged**